system/mp/test/mptest.cpp

```diff
@@ -521,127 +526,188 @@ void MultiTest(ICommunicator *_comm)
+        if ((k%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %u iteration %u complete", myrank, k);
+
+        if (k == iters)
+            break;
+    }
+    while (true);
 
-    int rnksumtotal = 0;
-    for(int i=0;i<numranks;i++)
-        rnksumtotal += (i+1);
+    PrintLog("MPTEST: MPRing complete");
 
-    PrintLog("MPTEST: MPITest myrank=%d numranks=%d rnksumtotal=%d", myrank, numranks, rnksumtotal);
+    mpicomm->barrier();
+
+    return;
+}
 
-    // send and recv to/from all others without a send/recv deadlock ...
+#define MSGLEN 1048576
 
-    mb.clear();
-    mb.append(myrank+1);
+void MPAlltoAll(IGroup *group, ICommunicator *mpicomm, size32_t buffsize=0, unsigned iters=0)
+{
+    rank_t myrank = group->rank();
+    rank_t numranks = group->ordinality();
+
+    if (numranks < 2)
+        throw MakeStringException(-1, "MPAlltoAll: MPRing Error, numranks (%u) must be > 1", numranks);
 
-    rank_t r;
-    int rankval;
-    int ranksum = myrank+1;
+    if (buffsize == 0)
+        buffsize = MSGLEN;
+    if (iters == 0)
+        iters = 1000;
+    if (iters < 1)
+        iters = 1;
 
-    int left, right;
+    PrintLog("MPTEST: MPAlltoAll myrank=%u numranks=%u buffsize=%u iters=%u", myrank, numranks, buffsize, iters);
 
-    if (numranks == 2)
+    // ---------
+
+    class Sender : public Thread
     {
-        if (myrank == 0)
+    public:
+        Linked<ICommunicator> mpicomm;
+        rank_t numranks;
+        rank_t myrank;
+        size32_t buffsize;
+        unsigned iters;
+        Sender(ICommunicator *_mpicomm, rank_t _numranks, rank_t _myrank, size32_t _buffsize, unsigned _iters) : mpicomm(_mpicomm), numranks(_numranks), myrank(_myrank), buffsize(_buffsize), iters(_iters)
         {
-            left = 1;
-            right = 1;
-            PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-            mpicomm->send(mb,right,MPTAG_TEST);
-
-            mb2.clear();
-            PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-            mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-            mb2.read(rankval);
-            ranksum += rankval;
         }
-        else
+
+        int run()
         {
-            left = 0;
-            right = 0;
-            mb2.clear();
-            PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-            mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-            mb2.read(rankval);
-            ranksum += rankval;
-
-            PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-            mpicomm->send(mb,right,MPTAG_TEST);
+            PrintLog("MPTEST: MPAlltoAll sender started, myrank = %u", myrank);
+
+            int pintvl = iters/10;
+            if (pintvl < 1)
+                pintvl = 1;
+
+            CMessageBuffer smb;
+            smb.appendBytes('a', buffsize);
+
+            for (unsigned k=1;k<=iters;k++)
+            {
+                bool oksend = mpicomm->send(smb, RANK_ALL_OTHER, MPTAG_TEST, MP_ASYNC_SEND);
```

Comment:
It doesn't matter, but suggestion of adding MP_ASYNC_SEND was in combination with doing away with a sender thread.

Reply:
ok, I will change it back.